Using a Factory to Create Services: Add working usage example.

BooleanType opened the pull request:
<!--

If your pull request fixes a BUG, use the oldest maintained branch that contains
the bug (see https://symfony.com/roadmap for the list of maintained branches).

If your pull request documents a NEW FEATURE, use the same Symfony branch where
the feature was introduced (and `master` for features of unreleased versions).

-->

Commit:
Add working usage example.

OskarStark commented on service_container/factories.rst:
```diff
@@ -293,5 +293,138 @@ previous examples takes the ``templating`` service as an argument:
         };
+        
+Usage example
+--------------------
```
Underline to long

BooleanType replied:
Thanks for advise, changes have been introduced.

OskarStark commented on service_container/factories.rst:
```diff
@@ -293,5 +293,138 @@ previous examples takes the ``templating`` service as an argument:
+    // src/Deliveries/DPD.php
+    namespace App\Deliveries;
+
+    class DPD
```
Let’s use DHL and UPS, it’s more  international

BooleanType replied:
Thanks for advise, changes have been introduced.

Commit:
Underline is too long; using DHL and UPS, it’s more international

Commit:
Morphological mistakes.

OskarStark commented on service_container/factories.rst:
```diff
@@ -293,5 +293,138 @@ previous examples takes the ``templating`` service as an argument:
+    // src/Deliveries/DHL.php
+    namespace App\Deliveries;
+
+    class DHL
```
To be more real life the should implement a marker interface.

BooleanType replied:
Thanks, done.

OskarStark commented on service_container/factories.rst:
```diff
@@ -293,5 +293,138 @@ previous examples takes the ``templating`` service as an argument:
+    
+    abstract class DeliveryFactory
+    {
+        public static function create($deliveryMethod)
```
The interface should be used here as typehint

BooleanType replied:
In case of my example I pass _the string_, which is name of subclass, but not the subclass object itself, 'cause factory creates the object based on passed string.
That's why I can't use typehint, but I've added typecheck inside `create()` instead:
 ```
$delivery = new $deliveryMethod;
if ( ! $delivery instanceof DeliveryInterface) {
    throw new RuntimeException(sprintf('%1$s should implement %2$s.', $deliveryMethod, DeliveryInterface::class));
}
```

OskarStark commented on service_container/factories.rst:
```diff
@@ -293,5 +293,138 @@ previous examples takes the ``templating`` service as an argument:
+            App\Deliveries\DHL:
+                factory:   ['@App\Factories\DeliveryFactory', create]
+                arguments:
+                    $deliveryMethod: 'App\Deliveries\DHL'
```
Let’s not use the variable key binding here, but:

arguments: [...]

Instead.

Same for xml and php

BooleanType replied:
Done.

Commit:
Marker interface; arguments.

Subclasses implement a marker interface.
Using arguments: [...] instead of the variable key binding in services.yaml.

OskarStark commented on service_container/factories.rst:
```diff
@@ -300,12 +300,23 @@ Usage example
+    interface DeliveryInterface
+    {
+    }
+
```
This can be removed completely

BooleanType replied:
@OskarStark, let me leave this until we reach a consensus on the use of the marker interface.

OskarStark commented on service_container/factories.rst:
```diff
@@ -330,12 +343,19 @@ Factory definition::
     
     abstract class DeliveryFactory
     {
         public static function create($deliveryMethod)
```
```suggestion
        public static function create(DeliveryInterface $deliveryMethod)
```
Afterwards the manuall check + exception can be removed

BooleanType replied:
@OskarStark, `$deliveryMethod` is not an object, but a _string_, which is passed in services.yaml, so it can't be typehinted with `DeliveryInterface ` (otherwise obvious exception is thrown):
```
pp\Deliveries\DHL:
      factory:   ['@App\Factories\DeliveryFactory', create]
      arguments: ['App\Deliveries\DHL']   #<-- $deliveryMethod is string 'App\Deliveries\DHL'
```
And _object_ is created basing on this string - this is an approach of factory method I've used.

Let me show you very simple code snippet I've found, which I used as a base for my example creation:

```
abstract class Animal
{
    // factory method
    public static function initial($animal)
    {
        // Object is created from string value.
        return new $animal();
    }
    abstract public function voice();
}

class Lion extends Animal
{
    public function voice()
    {
        echo 'I am a lion <br />' . PHP_EOL;
    }
}

class Cat extends Animal
{
    public function voice()
    {
        echo 'I am a cat <br />' . PHP_EOL;
    }
}

// As you see, string is passed to factory.
$animal1 = Animal::initial('Lion');
$animal2 = Animal::initial('Cat');

$animal1->voice();
$animal2->voice();
```

My example for factory in Symfony is just to show, how it works _with settings in services.yaml_ and without `implements` keyword. Every programmer will be able to run the example, but at the same time they also will be able to _modify_ it at their discretion.

Maybe, better approach for showing factory method exists. I preferred to use as simple working example as possible to emphasize how this works in Symfony.

Commit:
Return type declaration in factory

HeahDude commented on service_container/factories.rst:
```diff
@@ -293,5 +293,149 @@ previous examples takes the ``templating`` service as an argument:
+	
+    use App\Deliveries\DeliveryInterface;
+    
+    abstract class DeliveryFactory
```
An abstract class cannot be used as a factory.

BooleanType replied:
@HeahDude I'm afraid, I don't understand you... Matt Zandstra gives factory method's example with abstract class in his book. And example below works in Symfony without any exceptions. May I ask for the details of your comment?

HeahDude replied:
It currently works because the factory method happens to be static, but that example code is very fragile.

BooleanType replied:
Changed it in my today pull request.

Commit:
Factories usage example remake

Commit:
Fix "Please reorder the use statements alphabetically"

BooleanType commented:
Hello!
Sorry for soooo long pause. I'm still here :)
I've pushed some changes:
	- `DeliveryFactory` is not an abstract class anymore;
	- little refactoring has been made;
	- some mistakes are fixed.
PS. Sorry for that mess with other members commits history in my current pull request. My project version is out-of-date, but I didn't want to create _new_ pull request for 5.2 version for the same topic. So I decided to rebase everything, as it explained [here](https://symfony.com/doc/current/contributing/code/pull_requests.html#rebase-your-pull-request). Hope I've done it right. I still can close this PR for 5.0 and create new one for 5.2, if it's needed.

Commit:
Change controller method name for visual compability

BooleanType commented:
Hello-hello :) So what about my proposition to add this example into the documentation?

BooleanType commented:
@xabbuh

OskarStark commented:
As this is quite old and didn't get much attention, I am closing this one. Feel free to reopen against `6.4` branch. Thanks
